sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala

```diff
@@ -63,15 +63,26 @@ case class Concat(children: Seq[Expression]) extends Expression with ImplicitCas
 
   override protected def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
     val evals = children.map(_.genCode(ctx))
-    val inputs = evals.map { eval =>
-      s"${eval.isNull} ? null : ${eval.value}"
-    }.mkString(", ")
-    ev.copy(evals.map(_.code).mkString("\n") + s"""
-      boolean ${ev.isNull} = false;
-      UTF8String ${ev.value} = UTF8String.concat($inputs);
-      if (${ev.value} == null) {
-        ${ev.isNull} = true;
+    val argNums = evals.length
+    val args = ctx.freshName("argLen")
+    ctx.addMutableState("UTF8String[]", args, "")
+
+    val inputs = evals.zipWithIndex.map { case (eval, index) =>
+      if (eval.isNull != "true") {
```

Comment:
If `eval.isNull` is not a pre-evaluated constant?

Reply:
If `eval.isNull` is not a pre-evaluated constant, I expect that the following code at lines 73-74 will be generated. Only when we ensure it is `true`, we can avoid assigning a value (use default `null` value).